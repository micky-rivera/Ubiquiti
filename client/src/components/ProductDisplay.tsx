import React, { useEffect, useState } from 'react';
import ListItem from './ListItem';
import { useAppSelector, useAppDispatch } from '../hooks/hooks';
import { setList } from '../slices/slices'
import GridItem from './GridItem';

let allProducts: any;

function List() {
    const dispatch = useAppDispatch();
    const state = useAppSelector(store => store.app);
    const products = state.productList || [{name: '', line: '', deviceId: ''}];

    const parseData = (data: any) => {
        if (state.search.length > 0) {
            return data.filter((product: any) => product.name.toLowerCase().includes(state.search.toLowerCase()));
        }
        return data;
    }

    useEffect(()=> {
        fetch('http://localhost:8080/api/all')
        .then(res => res.json())
        .then(data => {
            allProducts = data;
            dispatch(setList(parseData(allProducts)));
        })
    }, [])

    useEffect(() => {
        dispatch(setList(parseData(allProducts)));
    }, [state.search])

    return (
        <>
            <div className={state.format === 'list' ? 'list' : 'grid'}>
                <div className={state.format === 'list' ? 'list-header' : 'hidden'}>
                    <p className='list-header__count'>{state.productList?.length || ''} devices</p>
                    <h4 className='list-header__line'>PRODUCT LINE</h4>
                    <h4 className='list-header__name'>NAME</h4>
                </div>
                <p className={state.format === 'grid' ? 'grid-header__count' : 'hidden'}>{state.productList?.length || ''} devices</p>
                {products.map((product, index) => (
                    <ListItem key={index} name={product.name} line={product.line} deviceId={product.deviceId} />
                ))}
                {products.map((product, index) => (
                    <GridItem key={index} name={product.name} line={product.line} deviceId={product.deviceId} />
                ))}
            </div>
        </>
    );
}

export default List;
