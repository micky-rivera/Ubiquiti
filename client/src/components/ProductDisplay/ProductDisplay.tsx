import React, { useEffect, useState } from 'react';
import ListItem from '../ListItem/ListItem';
import { useAppSelector, useAppDispatch } from '../../hooks/hooks';

let allProducts: Product[] = [];
const address = process.env.NODE_ENV === "development" ? "http://localhost:8080" : "";

function List() {
    const dispatch = useAppDispatch();
    const state = useAppSelector(store => store.app);
    const [products, setproducts] = useState<Product[]>([{name: '', line: '', deviceId: '06a25b40-ef1f-463a-82d9-13236866ea3d', details: []}]);
    const [notFound, setNotFound] = useState(false);

    const parseData = (data: Product[]) => {
        let searchTerm = '';
        let filters: string[] = [];
        if (window.localStorage.search) {
            searchTerm = JSON.parse(window.localStorage.search).term;
        }
        if (window.localStorage.filters) {
            filters = JSON.parse(window.localStorage.filters).terms;
        }
        if (searchTerm.length > 0 || filters.length > 0) {
            if (filters.length > 0) {
                return data.filter((product: Product) => product.name.toLowerCase().includes(searchTerm.toLowerCase()))
                .filter((product: Product) => filters.includes(product.line.toLowerCase())); // need to make this case insensitive!
            }
            return data.filter((product: Product) => product.name.toLowerCase().includes(searchTerm.toLowerCase()));
        }
        return data;
    }

    useEffect(()=> {
        fetch(`${address}/api/all`)
        .then(res => res.json())
        .then(data => {
            allProducts = data;
            setproducts(parseData(allProducts));
            setNotFound(false);
        })
        .catch((err) => {
            setNotFound(true);
        })
    }, [])

    useEffect(() => {
        setproducts(parseData(allProducts));
    }, [state.search])
    
    useEffect(() => {
        setproducts(parseData(allProducts));
    }, [state.filters])

    return (
        <>
            <div className={notFound ? 'hidden' : ''}>
                <div className={state.format === 'list' ? 'list' : 'grid'} data-testid='list'>
                    <div className={state.format === 'list' ? 'list-header' : 'hidden'}>
                        <p className='list-header__count'>{products.length || '0'} devices</p>
                        <h4 className='list-header__line'>PRODUCT LINE</h4>
                        <h4 className='list-header__name'>NAME</h4>
                    </div>
                    <p className={state.format === 'grid' ? 'grid-header__count' : 'hidden'}>{products.length || '0'} devices</p>
                    {products.map((product: Product, index) => (
                        <ListItem key={index} name={product.name} line={product.line} deviceId={product.deviceId} details={product.details} />
                    ))}
                </div>
            </div>
            <div className={notFound ? 'list__not-found' : 'hidden'}>
                <h2>Oops! There was an issue fetching the products.</h2>
                <h3>Please try again later.</h3>
            </div>
        </>
    );
}

export default List;
