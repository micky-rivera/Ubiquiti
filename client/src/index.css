@import './styles/header.css';
@import './styles/list.css';
@import './styles/nav.css';
@import './styles/grid.css';

* {
  margin: 0px;
  padding: 0px;
  border: 0px;
  box-sizing: border-box;
}

.hidden {
  display: none;
}

body {
  margin: 0;
  font-family: 'Lato', sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New',
    monospace;
}
