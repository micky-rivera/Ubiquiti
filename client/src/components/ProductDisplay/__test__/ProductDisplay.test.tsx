import React from "react";
import ProductDisplay from '../ProductDisplay';
import { Provider } from 'react-redux'
import {store} from '../../../store/store'
import { render, cleanup } from "@testing-library/react";
import TestRenderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom'
import { createRoot } from 'react-dom/client';

afterEach(cleanup);

it('renders without crashing', () => {
    const div = document.createElement('div');
    const root = createRoot(div);
    root.render(<BrowserRouter><Provider store={store}><ProductDisplay /></Provider></BrowserRouter>)
})

it('renders device count', () => {
    const { getByTestId } = render(<BrowserRouter><Provider store={store}><ProductDisplay /></Provider></BrowserRouter>);
    expect(getByTestId('list')).toHaveTextContent('devices');
})

it('renders the name header', () => {
    const { getByTestId } = render(<BrowserRouter><Provider store={store}><ProductDisplay /></Provider></BrowserRouter>);
    expect(getByTestId('list')).toHaveTextContent('NAME');
})

it('renders the line header', () => {
    const { getByTestId } = render(<BrowserRouter><Provider store={store}><ProductDisplay /></Provider></BrowserRouter>);
    expect(getByTestId('list')).toHaveTextContent('PRODUCT LINE');
})

it('matches snapshot', () => {
    const tree = TestRenderer.create(<BrowserRouter><Provider store={store}><ProductDisplay /></Provider></BrowserRouter>).toJSON();
    expect(tree).toMatchSnapshot();
})