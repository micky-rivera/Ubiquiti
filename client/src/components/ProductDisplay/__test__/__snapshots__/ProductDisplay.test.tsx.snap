// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
Array [
  <div
    className=""
  >
    <div
      className="list"
      data-testid="list"
    >
      <div
        className="list-header"
      >
        <p
          className="list-header__count"
        >
          1
           devices
        </p>
        <h4
          className="list-header__line"
        >
          PRODUCT LINE
        </h4>
        <h4
          className="list-header__name"
        >
          NAME
        </h4>
      </div>
      <p
        className="hidden"
      >
        1
         devices
      </p>
      <div
        className="list-item"
        data-testid="list-item"
        onClick={[Function]}
      >
        <div
          className="list-item__img"
        >
          <img
            data-testid="img"
            src="https://static.ui.com/fingerprint/ui/icons/06a25b40-ef1f-463a-82d9-13236866ea3d_25x25.png"
          />
        </div>
        <p
          className="list-item__line"
        />
        <p
          className="list-item__name"
        />
        <p
          className="hidden"
        />
        <p
          className="hidden"
        />
      </div>
    </div>
  </div>,
  <div
    className="hidden"
  >
    <h2>
      Oops! There was an issue fetching the products.
    </h2>
    <h3>
      Please try again later.
    </h3>
  </div>,
]
`;
